File: record_android/android/src/main/kotlin/com/llfbandit/record/record/recorder/engine/CaptureEngine.kt
package com.llfbandit.record.record.recorder.engine

import com.llfbandit.record.record.model.RecordConfig

const val DEFAULT_AMPLITUDE_DB = -160.0

/** Something the engine did on its own initiative; may be raised from any thread. */
sealed interface CaptureEvent {
  class Chunk(val bytes: ByteArray) : CaptureEvent

  /** Capture ended on its own; the engine has already released itself. */
  class Failed(val cause: Throwable) : CaptureEvent
}

/** Runs one recording end to end on the control thread; single-use, done after [stop]. */
interface CaptureEngine {
  /** Starts capturing; returns the effective (possibly renegotiated) config. */
  @Throws(Exception::class)
  fun start(): RecordConfig

  /** @return false if the engine cannot pause right now. */
  fun pause(): Boolean

  fun resume(): Boolean

  /** Releases everything, then answers with the capture or teardown error. */
  fun stop(delete: Boolean, done: (Throwable?) -> Unit)

  /** Latest input level in dB; [DEFAULT_AMPLITUDE_DB] while nothing is captured. */
  val amplitude: Double
}

File: record_ios/ios/record_ios/Sources/record_ios/core/CaptureEngine.swift
import Foundation

// Something the engine did by itself. It can come from any thread.
enum CaptureEvent {
  case chunk(Data)
  // Capture cannot continue. The controller ends the take.
  case terminated(Error?)
}

// Runs one take from start to end. Use it once: it is done after stop().
// The handler is set at init and never changes.
protocol CaptureEngine: AnyObject {
  // Starts to capture. Returns the config we really use, which may differ.
  func start() throws -> RecordConfig

  // Returns false if the engine cannot pause or resume now.
  func pause() -> Bool
  func resume() throws -> Bool

  // Frees everything. Returns the recorded file if there is one.
  @discardableResult
  func stop(delete: Bool) -> String?

  // Last input level in dB. It is `silenceDb` when we capture nothing.
  var amplitude: Float { get }
}
